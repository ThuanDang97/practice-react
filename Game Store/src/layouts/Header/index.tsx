import { Link, useNavigate } from 'react-router-dom'
import { Box, Drawer, IconButton, Typography, useTheme } from '@mui/material'
import { RiShoppingBag2Line } from 'react-icons/ri'
import PersonIcon from '@mui/icons-material/Person'
import { ChangeEvent, useMemo, useState } from 'react'
import ArrowForwardIcon from '@mui/icons-material/ArrowForward'

// Components
import SearchBar from '@components/SearchBar'
import Button from '@components/Button'
import CartItem from '@components/CartItem'

// Types
import { Game } from '@self-types/Game.types'

// Hooks
import { useAuthContext } from '@hooks/useAuthContext'
import { useCartContext } from '@hooks/useCartContext'

const Header = () => {
  const theme = useTheme()
  const navigate = useNavigate()
  const { userSession, logout } = useAuthContext()
  const { listCart, clearAll, deleteItem } = useCartContext()

  const username = useMemo(
    () =>
      userSession &&
      userSession?.user.firstName + ' ' + userSession?.user.lastName,
    [userSession],
  )

  const [valueSearch, setValueSearch] = useState('')
  const [open, setState] = useState(false)

  const cartQuantity = useMemo(() => listCart.length, [listCart.length])

  const totalPrice = useMemo(
    () =>
      listCart.reduce((acc: number, obj: { price: number }) => {
        return acc + obj.price
      }, 0),
    [listCart],
  )

  const toggleDrawer = () => {
    setState((prev) => !prev)
  }

  // Handle Logout
  const handleLogout = () => {
    logout()
  }

  // Handle login
  const handleClickButtonLogin = () => {
    navigate('/login')
  }

  // Handle change search
  const handleChangeSearch = (e: ChangeEvent<HTMLInputElement>) => {
    setValueSearch(e.target.value)
  }

  // Handle search when user press enter
  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (valueSearch !== '' && valueSearch !== undefined) {
      if (e.key === 'Enter') {
        navigate(`/games/search=${valueSearch}`)
      }
    } else {
      navigate('/')
    }
  }

  // Handle search when user press submit button
  const handleSubmitSearch = () => {
    if (valueSearch !== '' && valueSearch !== undefined) {
      navigate(`/games/search=${valueSearch}`)
    } else {
      navigate('/')
    }
  }

  // Clear Cart
  const handleClearCart = async () => {
    await clearAll(userSession?.user.id as string)
  }

  // Handler Deleted Item of Cart
  const handleDeletedItemCart = async (game: Game) => {
    const item = listCart.filter((gameItem) => game !== gameItem)

    await deleteItem(item, userSession?.user.id as string)
  }

  // Handle checkout
  const handleCheckout = () => {
    alert(
      `You will checkout with a product quantity of ${cartQuantity} and a total price of $${totalPrice.toFixed(
        2,
      )}`,
    )
  }

  return (
    <Box
      component="header"
      display="flex"
      alignItems="center"
      justifyContent="space-between"
      height={80}
      padding="0 50px"
      bgcolor={theme.palette.grey[900]}
    >
      <Link to="/">
        <Typography
          component="h1"
          sx={{ color: theme.palette.common.white }}
          fontSize={25}
          fontWeight={700}
          alignItems="center"
          display="flex"
        >
          <Box
            component="img"
            alt="logo"
            src="/images/logo.svg"
            width={20}
            height={20}
            marginRight={1}
          />
          Game Store
        </Typography>
      </Link>
      <SearchBar
        placeholder="Search games..."
        onKeyDown={handleKeyDown}
        onClick={handleSubmitSearch}
        onchange={handleChangeSearch}
      />
      <Box display="flex" alignItems="center" gap={2}>
        {username ? (
          <>
            <Button onClick={handleLogout} variant="contained">
              Logout
            </Button>
            <Typography
              display="flex"
              alignItems="center"
              color={theme.palette.common.white}
            >
              <PersonIcon fontSize="small" />
              {username}
            </Typography>
          </>
        ) : (
          <Button
            onClick={handleClickButtonLogin}
            color="success"
            variant="contained"
          >
            Login
          </Button>
        )}

        <IconButton
          onClick={toggleDrawer}
          size="small"
          sx={{
            color: theme.palette.common.white,
            display: 'flex',
            gap: 1,
            fontWeight: 700,
          }}
        >
          <RiShoppingBag2Line color={theme.palette.common.white} />
          <Typography color={theme.palette.common.white}>Cart</Typography>
          <Box
            bgcolor={theme.palette.common.white}
            borderRadius={12}
            width={20}
            color={theme.palette.common.black}
          >
            <Typography color={theme.palette.common.black}>
              {cartQuantity}
            </Typography>
          </Box>
        </IconButton>
      </Box>

      <Drawer anchor="right" open={open} onClose={toggleDrawer}>
        <Box
          sx={{
            p: 2,
            height: 1,
          }}
          bgcolor={theme.palette.grey[900]}
          width={450}
          position="relative"
        >
          <Box
            display="flex"
            justifyContent="space-between"
            alignItems="center"
            marginTop={2}
            padding="0 12px"
          >
            <Typography
              color={theme.palette.common.white}
              fontWeight={700}
              fontSize={30}
              variant="h3"
            >
              {Number(cartQuantity) !== 0
                ? `${cartQuantity} games`
                : 'No games added'}
            </Typography>
            {Number(cartQuantity) !== 0 && (
              <Button
                onClick={handleClearCart}
                variant="text"
                sx={{
                  fontSize: 18,
                  textTransform: 'capitalize',
                }}
              >
                Clear
              </Button>
            )}
          </Box>
          <Box
            display="flex"
            flexDirection="column"
            marginTop={2}
            alignItems="center"
            gap={1}
          >
            {listCart &&
              listCart.map((game, index) => (
                <CartItem
                  key={game.name + index.toString()}
                  game={game}
                  handleDelete={() => handleDeletedItemCart(game)}
                />
              ))}
          </Box>
          <Box
            display="flex"
            justifyContent="space-between"
            alignItems="center"
            bottom={10}
            position="absolute"
            width="95%"
          >
            <Typography
              color={theme.palette.grey[400]}
              fontWeight={600}
              fontSize={15}
            >
              Total: ${totalPrice.toFixed(2)}
            </Typography>
            <Button
              onClick={handleCheckout}
              variant="text"
              hovercolor={theme.palette.secondary.dark}
              sx={{
                color: theme.palette.common.white,
                fontSize: 24,
                textTransform: 'capitalize',
                fontWeight: 700,
              }}
            >
              Checkout
              <ArrowForwardIcon />
            </Button>
          </Box>
        </Box>
      </Drawer>
    </Box>
  )
}

export default Header
