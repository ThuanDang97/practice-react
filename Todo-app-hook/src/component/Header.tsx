import { useState } from 'react';

type TProps = {
  addTodo: (todo: string) => void;
};

const Header = (props: TProps): JSX.Element => {
  const [text, setText] = useState('');

  const handleGetValue = (e: React.ChangeEvent<HTMLInputElement>) => {
    setText(e.target.value);
  };

  const handleAddTodo = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter' && text) {
      props.addTodo(text);
      setText('');
    }
  };

  return (
    <header className="header">
      <h1>todos</h1>
      <input
        type="text"
        className="new-todo"
        placeholder="What needs to be done?"
        value={text}
        onChange={handleGetValue}
        onKeyPress={handleAddTodo}
      />
    </header>
  );
};
export default Header;
