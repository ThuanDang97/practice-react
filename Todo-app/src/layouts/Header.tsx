import React, { Component } from 'react';

type Props = {
  addTodo: Function;
};
class Header extends Component<Props> {
  state: {
    text: string;
  };
  constructor(props: Props) {
    super(props);
    this.state = {
      text: '',
    };
  }

  onAddTodo = (e: { key: string }) => {
    if (e.key === 'Enter' && this.state.text) {
      this.props.addTodo(this.state.text);
      this.setState({ text: (this.state.text = '') });
    }
  };

  render(): JSX.Element {
    return (
      <header className="header">
        <h1>todos</h1>
        <input
          type="text"
          className="new-todo"
          placeholder="What needs to be done?"
          value={this.state.text}
          onChange={(e) => this.setState({ text: e.target.value })}
          onKeyPress={(e) => this.onAddTodo(e)}
        />
      </header>
    );
  }
}

export default Header;
