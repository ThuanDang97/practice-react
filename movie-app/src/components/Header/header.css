header {
  display: flex;
  justify-content: space-around;
  align-items: center;
  top: 0;
  left: 0;
  background-image: radial-gradient(at 30% top, #031d33 0%, rgba(3, 37, 65, 1) 70%);
  height: 64px;
  width: 100%;
  z-index: 10;
  transition: top 0.2s linear;
}

.logo {
  margin-right: 16px;
  width: 154px;
  height: 20px;
}

nav ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}
