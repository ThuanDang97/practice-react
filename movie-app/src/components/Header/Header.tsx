import { Link } from 'react-router-dom';
import './header.css';
import logo from '../../assets/logo-blue-short.svg';
import { navHeader } from '../../constants/MockData';
import { Navbar } from '../Navbar/Navbar';

const Header = (): JSX.Element => {
  return (
    <header>
      <Link to="/">
        <h2 className="logo">
          <img src={logo} alt="" />
        </h2>
      </Link>
      <Navbar navItem={navHeader} />
    </header>
  );
};

export { Header };
