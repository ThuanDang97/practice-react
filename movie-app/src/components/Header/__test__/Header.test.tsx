import { render } from '@testing-library/react';
import { Header } from '../Header';
import { BrowserRouter as Router } from 'react-router-dom';

describe('Test Header component', () => {
  test('Component Header matches DOM Snapshot', () => {
    const headerComponent = (
      <Router>
        <Header />
      </Router>
    );

    const { container } = render(headerComponent);

    expect(container).toMatchSnapshot();
  });
});
