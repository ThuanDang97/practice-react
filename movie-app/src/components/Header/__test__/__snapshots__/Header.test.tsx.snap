// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Header component Component Header matches DOM Snapshot 1`] = `
<div>
  <header>
    <a
      href="/"
    >
      <h2
        class="logo"
      >
        <img
          alt=""
          src="IMAGE_MOCK"
        />
      </h2>
    </a>
    <nav
      data-testid="navbar"
    >
      <ul
        class="dropdown-menu"
      >
        <li
          class="dropdown-item"
        >
          <a
            href="/"
          >
            <span>
              Movies
            </span>
          </a>
          <div
            class="dropdown-content"
          >
            <a
              class="nav-link"
              data-testid="navitem"
              href="/"
            >
              <span>
                Popular
              </span>
            </a>
            <a
              class="nav-link"
              data-testid="navitem"
              href="/now-playing"
            >
              <span>
                Now Playing
              </span>
            </a>
            <a
              class="nav-link"
              data-testid="navitem"
              href="/up-coming"
            >
              <span>
                Upcoming
              </span>
            </a>
            <a
              class="nav-link"
              data-testid="navitem"
              href="/top-rated"
            >
              <span>
                Top Rated
              </span>
            </a>
          </div>
        </li>
        <li
          class="dropdown-item"
        >
          <a
            href="/favorite-movies"
          >
            <span>
              Favorite Movies
            </span>
          </a>
        </li>
      </ul>
    </nav>
  </header>
</div>
`;
